.github/workflows/build-and-test-differential.yaml

```diff
@@ -97,20 +97,28 @@ jobs:
         id: get-modified-packages
         uses: autowarefoundation/autoware-github-actions/get-modified-packages@v1
 
-      - name: Get changed files (existing files only)
-        id: get-changed-files
+      - name: Get required packages
+        id: get-required-packages
+        uses: mikefarah/yq@master
+        with:
+          cmd: yq '.[]' clang-tidy-required-packages.yaml
+
+      - name: Get target packages
+        id: get-target-packages
         run: |
-          echo "changed-files=$(git diff --name-only "origin/${{ github.base_ref }}"...HEAD | grep -E '\.(cpp|hpp)$' | while read -r file; do [ -e "$file" ] && echo -n "$file "; done)" >> $GITHUB_OUTPUT
+          echo "${{ steps.get-required-packages.outputs.result }}" > required_packages.txt
+          echo "${{ steps.get-modified-packages.outputs.modified-packages }}" | tr ' ' '\n' > modified_packages.txt
+          PACKAGES=$(grep -Fxf required_packages.txt modified_packages.txt | xargs)
+          echo "target-packages=${PACKAGES}" >> $GITHUB_OUTPUT
         shell: bash
 
       - name: Run clang-tidy
-        if: ${{ steps.get-changed-files.outputs.changed-files != '' }}
-        uses: autowarefoundation/autoware-github-actions/clang-tidy@v1
+        if: ${{ steps.get-target-packages.outputs.target-packages != '' }}
+        uses: yhisaki/autoware-github-actions/clang-tidy@use-ament_clang_tidy
         with:
           rosdistro: humble
-          target-packages: ${{ steps.get-modified-packages.outputs.modified-packages }}
-          target-files: ${{ steps.get-changed-files.outputs.changed-files }}
```

Comment:
Why transition from running it solely on the changed files to applying it across the entire modified packages?

Reply:
One reason is that clang-tidy does not work on a `.hpp` file.
In other words, `clang-tidy -p build/ hoge.hpp` will omit a lot of false positive warnings due to the lack of compilation infomation, and thus those .hpp files should be analyzed through .cpp files.

For example, you can compare the following two runs:

```
$ clang-tidy-14 --use-color -header-filter=src/universe/autoware.universe/.* -p ./build/ src/universe/autoware.universe/planning/sampling_based_planner/autoware_sampler_common/include/autoware_sampler_common/constraints/ma
p_constraints.hpp
```

```
$ export FILES=($(find src/universe/autoware.universe/planning/sampling_based_planner/autoware_sampler_common/ -name "*.cpp" -or -name "*.hpp"))
veqcc@dpc2206003:~/work/autoware$ run-clang-tidy -j 16 -p build/ "${FILES[@]}"

Reply:
Filtering the warnings could be done after clang-tidy ran, i.e. run on all files of the modified package, and then only report the errors/warnings that are in the modified files.

Reply:
@mojomex 
Thanks, I underdstand it :+1: 

The problem is that, when `changed-files` contains `.hpp` files, then clang-tidy should be run on `.cpp` files who include those hpp files.
To realize it, we want to analyze all the files in the target package. That is one readon why we want to introduce the `clang-tidy-required-packages.yaml` filtering.

Reply:
@veqcc Ah, sorry, I expressed my idea a bit loosely!

I understand that Clang-Tidy has to run on all files, including (possibly unmodified) `.cpp` files.
My proposal would be to

1. run on all files of the _modified package_ (or, if it needs to run on all of Autoware, that is also fine I guess)
2. Filter the warnings/errors from Clang-Tidy and only report those in the _modified files_

This could be done via the `--line-filter` argument of clang-tidy and passing the _modified files_ to it:
```
  --line-filter=<string>         - List of files with line ranges to filter the
                                   warnings. Can be used together with
                                   -header-filter. The format of the list is a
                                   JSON array of objects:
                                     [
                                       {"name":"file1.cpp","lines":[[1,3],[5,7]]},
                                       {"name":"file2.h"}
                                     ]
```

I'll have to try that when I have time :slightly_smiling_face:

Reply:
@mojomex 
> Filter the warnings/errors from Clang-Tidy and only report those in the modified files

I agree to the idea that it's better to refine errors/warnings to the modified files.

However, I am not sure whether the changes in `.hpp` files never invoke a new error/warning in `.cpp` files who include it.
In case of it, I want to include all the errors/warnings in CI results.

Reply:
It makes sense, thanks for explaining!